Add stm to docs matrix

hierophect opened the pull request:
As per title. Removes a number of build flags as redundant and reworks the stm port `mpconfigport.mk` to match. Adjusts `shared_bindings_matrix.py` to recognize STM32 family categories. Adds a minor addition to porting.rst to explain how build flags impact porting. 

Builds locally without issue, and correctly lists modules for all STM32 boards in the support matrix list, including the thunderpack (which enables NVM) and the H7 and F7 (which have restricted modules). 

Resolves #2587 
Resolves #2458

sommersoft commented:
I meant to add: I'll hold off on merging. I'm not a huge fan of sending known-incorrect documentation out. Without a better grasp on what is causing the noted problem, I'm not sure how a long a fix would take.

Thoughts on holding this until I (or whomever) gets a handle on the secondary problem?

dhalbert commented:
Are there similar issues on other ports, or just stm32f4?

sommersoft commented:
@dhalbert, looking at samples for different ports, I'm not seeing similar issues. 

I'm pretty confident that its only related to the use of [`CIRCUITPY_MINIMAL_BUILD`](https://github.com/adafruit/circuitpython/blob/master/ports/stm32f4/mpconfigport.mk#L16). The minimal build logic was added after `docs/shared_bindings_matrix.py` was written, so it just doesn't handle it.

dhalbert commented:
OK, I will add a 'changes requested' review to hold off on this until the `.py` script is fixed.

hierophect commented:
@dhalbert do you think it'd be worth it for me to address #2458 as a part of this? Just do it all in one go?

dhalbert commented:
@hierophect I think so, if you can get it to work, since then the doc will be correct from the start.

hierophect commented:
@sommersoft I've adjusted the shared_bindings_matrix.py file - builds ok for me but let me know if you'd like to see any changes. 

@jepler could you double check the change I made to `MICROPY_COMP_FSTRING_LITERAL` in circuitpy_mpconfig.h? I removed MINIMAL_BUILD as a flag - are there still conditions where you'd like this to be optional?

jepler commented on py/circuitpy_mpconfig.h:
```diff
@@ -179,17 +179,11 @@ typedef long mp_off_t;
-	#define MICROPY_CPYTHON_COMPAT                (CIRCUITPY_FULL_BUILD)
-#endif
-#if !defined(MICROPY_COMP_FSTRING_LITERAL)
+#define MICROPY_CPYTHON_COMPAT                (CIRCUITPY_FULL_BUILD)
 #define MICROPY_COMP_FSTRING_LITERAL          (MICROPY_CPYTHON_COMPAT)
-#endif
```
I didn't find any evidence that MICROPY_COMP_FSTRING_LITERAL was presently defined before reaching this line; the compiler should shout if it was.  If it's accepted, and if no boards suddenly overflow due to fstring literals being compiled in, it's probably fine.

hierophect commented:
@sommersoft I liked the concept of ALWAYS/DEFAULT/MINIMAL at first, but I realized it was backwards in how it affected ports. Setting everything to off and then adding what you support meant that you had to memorize what was in each "level" and then constantly re-evaluate if you should be moving up to the next. Starting at 100% and having a list of everything you turn off is like having a built in TODO list, one that you continually whittle down at. So I ultimately think it's better this way, with no categories at all except the one that SAMD21G18s need to restrict flash size.

hierophect commented:
@dhalbert could you give this another look?

dhalbert commented on docs/porting.rst:
```diff
@@ -42,6 +42,43 @@ to the port's directory (in the top level until the ``ports`` directory is
+    CIRCUITPY_I2CSLAVE = 0
+    CIRCUITPY_DISPLAYIO = 0       # Requires SPI, PulseIO (stub ok)
+
+    # Modules with no common-hal implementation, but depend on something else
+    CIRCUITPY_BITBANGIO = 0       # Requires DigitalIO
+    CIRCUITPY_GAMEPAD = 0         # Requires DigitalIO
+    CIRCUITPY_PIXELBUF = 0        # Does nothing without a neopixel or dotstar
+    CIRCUITPY_RANDOM = 0          # Requires OS
+    CIRCUITPY_STORAGE = 0         # Requires OS, filesystem
+    CIRCUITPY_TOUCHIO = 0         # Requires Microcontroller
+    CIRCUITPY_USB_HID = 0         # Requires USB
+    CIRCUITPY_USB_MIDI = 0        # Requires USB
+    CIRCUITPY_REQUIRE_I2C_PULLUPS = 0 # Does nothing without I2C
+    CIRCUITPY_ULAB = 0            # No requirements, but takes extra flash
```
These are implemented in `shared-module`, and are port-independent, so as soon as their port-dependent dependencies are implemented, they can be turned on. So could you change the comment to reflect that? For `CIRCUITPY_PIXELBUF`, it doesn't have dependencies, but it's not useful until either `neopixel_write` (NeoPixels) or `SPI` (DotStars) are available.

hierophect replied:
Good idea.

dhalbert commented on docs/shared_bindings_matrix.py:
```diff
@@ -140,6 +140,9 @@ def get_excluded_boards(base):
             chip_keyword = "CHIP_FAMILY"
         elif port in ["nrf"]:
             re_board_chip = re.compile("MCU_VARIANT\s=\s(\w+)")
```
This should be a raw string: `r"..."`.

dhalbert commented on docs/shared_bindings_matrix.py:
```diff
@@ -140,6 +140,9 @@ def get_excluded_boards(base):
         elif port in ["nrf"]:
             re_board_chip = re.compile("MCU_VARIANT\s=\s(\w+)")
+        elif port in ["stm"]:
+            re_board_chip = re.compile("MCU_SERIES\s*=\s*(\w+)")
```
Should be raw string.

dhalbert commented on ports/stm/mpconfigport.mk:
```diff
@@ -1,106 +1,56 @@
-	CIRCUITPY_BOARD = 1
-	CIRCUITPY_DIGITALIO = 1
+ifeq ($(MCU_SERIES),F4)
+	# Items requiring common-hal implementation:
```
Do you mean "not yet implemented"?

hierophect replied:
I meant that I need to create a common-hal folder and implementation for them, but I can switch it around if it seems unclear.

dhalbert commented on ports/stm/mpconfigport.mk:
```diff
@@ -1,106 +1,56 @@
+endif
+
+ifeq ($(MCU_SERIES),H7)
+	# Items requiring common-hal implementation:
```
ditto

dhalbert commented on ports/stm/mpconfigport.mk:
```diff
@@ -1,106 +1,56 @@
+endif
+
+ifeq ($(MCU_SERIES),F7)
+	# Items requiring common-hal implementation:
```
ditto

dhalbert commented on py/circuitpy_mpconfig.h:
```diff
@@ -179,17 +179,11 @@ typedef long mp_off_t;
 
-// CIRCUITPY_FULL_BUILD is defined in a *.mk file.
-
+// CIRCUITPY_FULL_BUILD is defined in py/circuitpy_mpconfig.mk
```
The point of the original (not well written) comment is that the default value of `CIRCUITPY_FULL_BUILD` could be overriden in a .mk file. Right now there's no example of that.

dhalbert replied:
You could just remove the comment now.

hierophect replied:
I honestly think the whole system is a little dangerous, because it has two separate flags that do the same thing (SMALL_BUILD = !FULL_BUILD) and you could easily do something like setting FULL_BUILD=0, SMALL_BUILD=0 without realizing. It might be better if we remove SMALL_BUILD altogether, but I didn't want to go overboard.

dhalbert replied:
`CIRCUITPY_SMALL_BUILD` is used in `atmel-samd` for 192kB builds (and only in that port). So it could probably be removed from here, but something equivalent still needs to be in `atmel-samd`.

hierophect replied:
Right, what I'm saying is that it might be better if atmel-samd 192kB builds did this by setting CIRCUITPY_FULL_BUILD=0, rather than using SMALL_BUILD which literally just does the same thing but with an extra step.

dhalbert commented on py/circuitpy_mpconfig.mk:
```diff
@@ -222,7 +190,7 @@ endif
 CFLAGS += -DCIRCUITPY_PROTOMATTER=$(CIRCUITPY_PROTOMATTER)
```
This is now called `CIRCUITPY_RGBMATRIX`.

hierophect replied:
Was this introduced in the last few days? I remember you mentioning it but I thought I pulled right before submitting this PR...

dhalbert replied:
I looked in the latest master, because I thought it was merged already. So maybe try fetch and merge again.

dhalbert commented on py/circuitpy_mpconfig.mk:
```diff
@@ -39,38 +38,8 @@ ifndef CIRCUITPY_FULL_BUILD
 endif
 CFLAGS += -DCIRCUITPY_FULL_BUILD=$(CIRCUITPY_FULL_BUILD)
 
-# Setting CIRCUITPY_MINIMAL_BUILD = 1 will disable all features
-# Use for for early stage or highly restricted ports
-ifndef CIRCUITPY_MINIMAL_BUILD
-CIRCUITPY_MINIMAL_BUILD = 0
-endif
-CFLAGS += -DCIRCUITPY_MINIMAL_BUILD=$(CIRCUITPY_MINIMAL_BUILD)
-
-ifndef CIRCUITPY_DEFAULT_BUILD
-  ifeq ($(CIRCUITPY_MINIMAL_BUILD),1)
-        CIRCUITPY_FULL_BUILD = 0
-        CIRCUITPY_DEFAULT_BUILD = 0
-  else
-        CIRCUITPY_DEFAULT_BUILD = 1
-  endif
-endif
-CFLAGS += -DCIRCUITPY_DEFAULT_BUILD=$(CIRCUITPY_DEFAULT_BUILD)
-
-# Some features have no unique HAL component, and thus there's never
-# a reason to not include them.
-ifndef CIRCUITPY_ALWAYS_BUILD
-  CIRCUITPY_ALWAYS_BUILD = 1
-endif
-CFLAGS += -DCIRCUITPY_ALWAYS_BUILD=$(CIRCUITPY_ALWAYS_BUILD)
-
-
-# All builtin modules are listed below, with default values (0 for off, 1 for on)
-# Some are always on, some are always off, and some depend on CIRCUITPY_FULL_BUILD.
-#
-# *** You can override any of the defaults by defining them in your mpconfigboard.mk.
-
```
Did you write this originally, or did Scott or I? I can't remember.

hierophect replied:
It was me, right when I got started. As I mentioned in my reply to Sommersoft, I got stuck on thinking about the problem backwards and wanted ports to build up from things being disabled, rather than the other way around. Nobody else uses it so I think it's best deleted.

sommersoft commented:
@dhalbert the raw strings were part of the cleanup I mentioned in my review. Nothing wrong with fixing them here; I just figured I'd clean up my own messes. 😄

hierophect commented:
@dhalbert ok fixes are in and merge completed. I went ahead and deleted SMALL_BUILD completely across the port, atmel simply sets FULL_BUILD=0 now.

dhalbert commented on ports/atmel-samd/mpconfigport.mk:
```diff
@@ -65,19 +65,19 @@ CIRCUITPY_SAMD = 1
 endif
 
 ifndef CIRCUITPY_ULAB
-ifneq ($(CIRCUITPY_SMALL_BUILD),1)
+ifneq ($(CIRCUITPY_FULL_BUILD),0)
 CIRCUITPY_ULAB = 1
 endif
 endif
```
This is redundant with what's already in `circuitpy_mpconfig.mk`, I think.

dhalbert commented on ports/atmel-samd/mpconfigport.mk:
```diff
@@ -65,19 +65,19 @@ CIRCUITPY_SAMD = 1
 
 ifndef CIRCUITPY_RGBMATRIX
-ifneq ($(CIRCUITPY_SMALL_BUILD),1)
+ifneq ($(CIRCUITPY_FULL_BUILD),0)
 CIRCUITPY_RGBMATRIX = 1
 endif
```
This is the same as `CIRCUITPY_RGBMATRIX = $(CIRCUITPY_FULL_BUILD)`

hierophect replied:
RGBMatrix is only enabled for the SAMD, so this is appropriate here. If it was full build, everything else would need to turn it off.

dhalbert replied:
I was just saying that the code
```
ifneq ($(CIRCUITPY_FULL_BUILD),0)
CIRCUITPY_RGBMATRIX = 1
endif
```
is equivalent to the shorter
```
CIRCUITPY_RGBMATRIX = $(CIRCUITPY_FULL_BUILD)
```

dhalbert replied:
CIRCUITPY_FULL_BUILD is either 0 or 1, so you can just set the variable to its value.

hierophect replied:
Oh you mean just here in this file, I see, it was the other way because it was setting the opposite of SMALL_BUILD but now that it's FULL_BUILD it's just one line. Changed.

hierophect replied:
Sorry, I thought you meant you wanted it relocated.

dhalbert commented on ports/atmel-samd/mpconfigport.mk:
```diff
@@ -65,19 +65,19 @@ CIRCUITPY_SAMD = 1
 
 ifndef CIRCUITPY_FRAMEBUFFERIO
-ifneq ($(CIRCUITPY_SMALL_BUILD),1)
+ifneq ($(CIRCUITPY_FULL_BUILD),0)
 CIRCUITPY_FRAMEBUFFERIO = 1
```
ditto

hierophect replied:
Ditto for RGBMatrix, this is labeled as though it's a SAMD only feature. is it?

hierophect replied:
Chatted with @jepler about this on Discord to verify that FrameBufferIO is only used with RGBMatrix on the SAMD for now. Both RGBMatrix or FrameBufferIO default to 0 right now (unlike ulab, which defaults on) so they should be defined port by port - I imagine we'll make them default to 1 sometime later once they have more coverage (doesn't seem like we have a specific line for that but I'd figure it's when all the "big" ports atmel-samd, nrf, mimxrt, and stm are supported).

hierophect commented:
@dhalbert should be all wrapped up for one last review!
